packages/netlify-cms-widget-relation/src/RelationControl.js

```diff
@@ -1,192 +1,189 @@
-    isFetching: PropTypes.bool,
+    field: ImmutablePropTypes.map,
     fetchID: PropTypes.string,
     query: PropTypes.func.isRequired,
-    clearSearch: PropTypes.func.isRequired,
     queryHits: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
     classNameWrapper: PropTypes.string.isRequired,
     setActiveStyle: PropTypes.func.isRequired,
     setInactiveStyle: PropTypes.func.isRequired,
   };
 
-  static defaultProps = {
-    value: '',
-  };
-
-  constructor(props, ctx) {
-    super(props, ctx);
-    this.controlID = uuid();
-    this.didInitialSearch = false;
-  }
-
-  componentDidMount() {
-    const { value, field } = this.props;
-    if (value) {
-      const collection = field.get('collection');
-      const searchFields = field.get('searchFields').toJS();
-      this.props.query(this.controlID, collection, searchFields, value);
-    }
+  shouldComponentUpdate(nextProps) {
+    return (
+      this.props.value !== nextProps.value ||
+      this.props.hasActiveStyle !== nextProps.hasActiveStyle ||
+      this.props.queryHits !== nextProps.queryHits ||
+      this.props.metadata !== nextProps.metadata
+    );
   }
 
   componentDidUpdate(prevProps) {
     /**
      * Load extra post data into the store after first query.
      */
     if (this.didInitialSearch) return;
-    if (
-      this.props.queryHits !== prevProps.queryHits &&
-      this.props.queryHits.get &&
-      this.props.queryHits.get(this.controlID)
-    ) {
+    const { value, field, forID, queryHits, onChange } = this.props;
+
+    if (queryHits !== prevProps.queryHits && queryHits.get(forID)) {
       this.didInitialSearch = true;
-      const suggestion = this.props.queryHits.get(this.controlID);
-      if (suggestion && suggestion.length === 1) {
-        const val = this.getSuggestionValue(suggestion[0]);
-        this.props.onChange(val, {
-          [this.props.field.get('name')]: {
-            [this.props.field.get('collection')]: { [val]: suggestion[0].data },
-          },
+      const valueField = field.get('valueField');
+      const hits = queryHits.get(forID);
+      if (value) {
+        const listValue = List.isList(value) ? value : List([value]);
+        listValue.forEach(val => {
+          const hit = hits.find(i => i.data[valueField] === val);
+          if (hit) {
+            onChange(value, {
+              [field.get('name')]: {
+                [field.get('collection')]: { [val]: hit.data },
+              },
+            });
+          }
         });
       }
     }
   }
 
-  onChange = (event, { newValue }) => {
-    this.props.onChange(newValue);
+  handleChange = selectedOption => {
+    const { onChange, field } = this.props;
+    let value;
+
+    if (Array.isArray(selectedOption)) {
+      value = selectedOption.map(optionToString);
+      onChange(fromJS(value), {
+        [field.get('name')]: {
+          [field.get('collection')]: {
+            [last(value)]: !isEmpty(selectedOption) && last(selectedOption).data,
```

Comment:
I'm not sure how the underlying mechanics works here but wouldn't it be better to have this as a list somehow – instead of just picking the last item

Reply:
The last item will always be the newly added item on the list, so we guaranteed that all selected item will have their metadata in store. We cannot use list here because that would mean using `List()` as key here  `[last(value)]`.